### ISSUE TITLE
Types are not tagged as types

### ISSUE THREAD

**Issue by matter123** [jeff-hykin/cpp-textmate-grammar#138]
This is a meta issue similar to #81 but for types.

Progress:
* :white_check_mark:  in structs/classes/enums/unions
* :running_man: in declarations #71
* :running_man: in parameters #134 
* :white_circle: after new expressions #129
* :question: in typedefs
* :question: using aliases

Key | Explanation
-----|----
:white_check_mark: | complete
:running_man: | in progress
:white_circle: | not started
:question: | has not been investigated
:large_blue_circle: | blocked (see issue)

**Comment by matter123** [comment]
@alekseyt Moving the discussion from #198 to here

So v1.9.0 Added support for tagging the return type. This currently causes inconsistencies because variable declarations do not have their type tagged. The current issue with tagging variable declaration types is that in `A * B;` it is unknown if `A *` is a type or if `A` and `B` are both variables multiplied together.

**Comment by alekseyt** [comment]
IMO the issue with ambiguity of expression is a big issue, but another issue is that there is no way to remove such inconsistencies or unwanted highlighting. It would be much easier for me personally just to remove highlighting in places where it's inconsistent.

The issue with inconsistency in highlighting is that i depend on it when i'm typing. If i type something and it's highlighted then i assume i typed it correctly without re-reading typed text. If i typed something and it isn't highlighted when i expect it to be highlighted then i need to stop and find a potential typo.

When highlighting is inconsistent then i can't have "muscle memory" for such things and every occurrence of "typo" is effectively false positive so to say. So it's better not to have highlighting at all rather than to have inconsistent highlighting, in general. It doesn't look very good too. This is just my personal opinion, just to explain my way of thinking.

So one way to fix this might be to add tags where they are missing and dive into issues with ambiguity, but another way might be to provide some way to just disable highlighting where it's unwanted, inconsistent or ambiguous.

For example, if `std::string` would be just no type always without going into `entity.name.scope-resolution` and other bits, or if `something::something()` would always be a function call, IMO it would be easier to style it and it could play better with other themes too without the need to tailor them for C++ specifically.

Just saying that instead of making things more complex it might be easier to make things simpler. I don't have a proof of that concept, just an idea. This is an issue for discussion, right?

But ability to style a scope name is a good thing too, i'm not in any way discouraging you from removing liberties from people who do C++ syntax highlighting specifically, it's a good stuff. Although there might not be a lot of people who would prefer to make a theme specifically for C++, it's often necessity to work with other languages at the same time.

Please reconsider this comment https://github.com/jeff-hykin/cpp-textmate-grammar/issues/59#issuecomment-481545491: "It seems really weird to not have the fix be the default". It would be really nice to have good defaults and it also would be really nice to have the ability to tailor theme for C++ too.

**Comment by alekseyt** [comment]
Consider this image:

![image](https://user-images.githubusercontent.com/993454/58375283-d5f50e00-7f57-11e9-965c-8b72404a3808.png)

This is how it would look like in Dark+ in vscode:

![image](https://user-images.githubusercontent.com/993454/58375293-063cac80-7f58-11e9-99ed-b9c2b03d624d.png)

I guess Dark+ is more consistent. Preferable approach for me here is to actually disable highlighting on `MyClass` in `MyClass::f() {}` to make it more consistent. So ultimately Dark+ approach wins just by having less sophisticated highlighting and not trying to tailor itself for C++ syntax, right?

**Comment by matter123** [comment]
entity.name.{function, type, namespace,etc} not extending to include the scope resolution is probably a bug and resolving that would help some of the inconsistencies.

The problem with simpler is that C++ is a complex language. Simple wither doesn't mark things at all or gets it wrong.

**Comment by alekseyt** [comment]
![image](https://user-images.githubusercontent.com/993454/58375452-0b9bf600-7f5c-11e9-8701-4a7948c4bdaf.png)

I don't say that liberty to highlight something is a bad thing. I'm just saying that it might be easier to disable some highlighting when it is needed for one reason or another. Just my 2c.

**Comment by matter123** [comment]
What is wrong with the highlighting of `MyClasd`? If it's the fact that it's a typo. The grammar has no way of knowing that.

We are working on highlighting `MyClass::MyClass()` see #80 (currently blocked by #127). So that should be improved.

**Comment by alekseyt** [comment]
There is nothing terribly wrong in tagging, i'm just trying to explain what i have to deal with when working on theme. You are absolutely right that C++ is a complex language and i'm not saying that syntaxing has to implement entire C++ compiler nor it has to be adjusted to fit particular theme.

All i'm saying is that it would be easier if namespace would be namespace everywhere and a function would be a function everywhere. If you have your own considerations - you are very welcome to have them. I hope my explanations did help at least for a bit, thank you for your hard work, i'm looking forward for future releases.
